src/pages/ApiPage.jsx

```diff
@@ -0,0 +1,116 @@
+    const initialState = subscribes.reduce((initialState, prop) => ({ [prop]: '', ...initialState }), baseInitialState);
+
+    const [state, setState] = useState(initialState);
+    const updateState = (changes) => setState({ ...state, ...changes });
+    
+    const reset = () => updateState(initialState);
+    const resetPrint = () => updateState({ request: {}, response: {} });
+    
+    const requestPerformer = (provider, reducer) => async () => {
+        // Check if provider.args is an array and if so, apply fn arguments from state.
+        const request = (Array.isArray(provider.args) && provider.args.length) ?
+            provider.fn.apply(null, provider.args.map((arg) => state[arg] || null)) :
+            provider.fn();
+    
+        updateState({ request });
+
+        const response = await API.makeRequest(stringifyBody(request));
+        updateState(reducer({ request, response }));
+    };
+
+    const performInitRequest = requestPerformer(
+        initializer,
+        isFunction(initializer.reducer) ? initializer.reducer : defaultInitReducer
+    );
+
+    const performSubmitRequest = requestPerformer(
+        submitter,
+        isFunction(submitter.reducer) ? submitter.reducer : defaultSubmitReducer
+    );
+
+    const initRequest = async () => {
+        reset();
+        performInitRequest();
+    };
+    
+    const submitRequest = async () => {
+        resetPrint();
+        await performSubmitRequest();
+        executeIfExists(submitCallback);
+    };
+    
+    const flowData = Object.entries({...props, ...state})
+        .filter(([key, _]) => flowDataKeys.includes(key))
+        .map(([key, value]) => ({
+            name: splitOnCapitals(key).map(capitalize).join(' '),
+            value,
+        }));
+
+    const subscribedState = subscribes.reduce(
+        (acc, stateVar) => ({ 
+            ...acc,
+            [`${stateVar}`]: hasKey(state, stateVar) ? state[stateVar] : null,
+            [`set${capitalize(stateVar)}`]: (newVal) => updateState({ [`${stateVar}`]: newVal }),
+        }), {});
+
+    return (
+        <Container>
+            <Row>
+                <Col md={{ span: 4 }}>
+                    <Component 
+                        {...props}
+                        initHandler={initRequest} 
+                        submitHandler={submitRequest} 
+                        {...subscribedState}    
+                    />
+                    <FlowData data={flowData} />
+                </Col>
+                <Col md={{ span: 8 }}>
+                    <DisplayRequest title="Request" text={state.request} />
+                    <DisplayRequest title="Reponse" text={state.response} />
+                </Col>
+            </Row>
+        </Container>
+    );
+};
+
+export const defaultInitReducer = ({ request, response }) => {
+    try {
+        const flowID = response.json.id;
+        const csrf = response.json.ui.nodes
+            .filter(x => x.attributes.name == "csrf_token")
```

Comment:
Export as const?

Reply:
Yeah. Thats by intention and actually used. A reducer returns a plain object. So if you want to not override the default reducer, but instead extend it's output, you can mix the exported default reducers in

```js
import { defaultInitReducer } from 'ApiPage';
// [...]

initializer: {
    fn: someRequestGenerator,
    reducer: ({request, reponse}} => ({ 
        ...defaultInitReducer({request, response}),
        foo: response.bar
    }),
};
```

Reply:
Here, the `defaultSubmitReducer` is used by the RegPage for that purpose.
[Line 23](https://github.com/mindtastic/stagefright/blob/e778a853cc41fadb9f2c6f9ba6636d8e83d2ad09/src/pages/RegPage.jsx).

Reply:
I mean just the csrf_token string